### PR TITLE
feature: move-to-space: Apiserver part

### PR DESCRIPTION
**Pull request by nammn**
### Checklist

 - [ ] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [ ] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?
 - [ ] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

- adds the apiserver part to move a list of CIDRs to a space from their old space.
- the cmd part will be implemented in another patch. The command will look like this:
```
juju move-to-space <spacename> <sliceOfCIDRs>
```

- **Additionally, this patch does not include the check and movement in case of endpoint bindings which are used and may or may not be moved as well. This work will be added in another patch.**

#### Possible cases which the code currently handles:
- [ ] Wrong CIDR (2)
- [ ] CIDR/Subnet does not exist (3)
- [ ] Space does not exist (4)
- [ ] CIDR is already on space (5)
- [ ] (1) but with force option (6)
- [ ] Success case (7)
- [ ] Moving a CIDR of a used machine to another space, which has a negative constraint. [8]
- [ ] Machine has an address on that subnet, but no constraint is blocking. [9]

## QA steps
There are no specific QA steps doable for now, as the CMD part is missing.
If someone wants to try the changes locally one can apply checkout the branch: https://github.com/nammn/juju/tree/updatespace-with-cmd

With this, there should be a basic cmd implementation to the move-to-space implementation 

[2] wrong CIDR
```
❯ juju move-to-space dsadas 132.31.0.32130/28
ERROR invalid arguments specified: "132.31.0.32130/28" is not a valid CIDR
```
[3] subnet not found
```
~
❯ juju move-to-space dsadas 132.31.0.0/28
ERROR cannot update space "dsadas": subnet "132.31.0.0/28" not found
```

[4] space not found
```
~
❯ juju move-to-space dsadas 172.31.0.0/28
ERROR cannot update space "dsadas": space "dsadas" not found
```

[5] already in space
```
~
❯ juju move-to-space db 172.31.0.0/28 
ERROR cannot update space "db": supplied CIDR "172.31.0.0/28" is already in space "db"
```

[7] basic success case
```
❯ juju deploy mediawiki 
~
❯ juju move-to-space db 172.31.0.0/28
Subnet "172.31.0.0/28" moved from "alpha" to "db"
```
<- note that 172.31.0.0/28 is not used on any machine 

[8] is more complex, I will add this as a step QA step here:
```
❯ juju spaces
Name   Space ID  Subnets       
alpha  0         172.31.0.0/28 
                 172.31.1.0/28 
                 172.31.16.0/20
                 172.31.2.0/28 
                 172.31.3.0/24 
db     1  
❯ juju deploy mediawiki --constraints spaces=alpha,^db
❯ juju move-to-space db 172.31.16.0/20
ERROR cannot update space "db": cannot move CIDR "172.31.16.0/20" from space "alpha" to space: "^db", as this would violate the current application constraint: "^db" on application "mediawiki".
```

[9] subnet in use, but no constraint

`juju deploy mysql`
... wait until machine, which mysql get deployed to, is online. In this case, it is machine "1"
`juju show-machine 1 `
--> look for the ip address matching a CIDR from a space. In my example, it is 172.31.26.60
```
❯ juju move-to-space db 172.31.16.0/20
Subnet "172.31.16.0/20" moved from "alpha" to "db"
```

## Spec
https://docs.google.com/document/d/1tE5GMF9Uw8W7QQoybgA_vs_RqCHQdbKXmQT_Kaaj6Pk/edit#heading=h.fro54rq52v4

### CODE REVIEW

File: api/spaces/spaces.go [hunk 1]
```diff
@@ -175,3 +175,7 @@ func (api *API) RenameSpace(oldName string, newName string) error {
 	}
 	return nil
 }
+
+func (api *API) UpdateSpace(name string, ids []string) error {
```
Comment (nammn) [inline api/spaces/spaces.go]: this will be finished in another patch

File: state/subnets.go [hunk 2]
```diff
@@ -252,42 +252,51 @@ func (s *Subnet) Update(args network.SubnetInfo) error {
 				return nil, errors.Trace(err)
 			}
 		}
-		makeSpaceNameUpdate, err := s.updateSpaceName(args.SpaceName)
+		return s.UpdateOps(args)
+	}
+	return errors.Trace(s.st.db().Run(buildTxn))
+}
+
+// UpdateOps returns the OPs to add new info to the subnet based on input info.
```
Comment (nammn) [inline state/subnets.go]: This part is only basic refactoring. No functional changes.

File: state/subnets.go [hunk 3]
```diff
@@ -252,42 +252,51 @@ func (s *Subnet) Update(args network.SubnetInfo) error {
...
+// Currently no data is changed unless it is the "undefined" space from MAAS.
+// There are restrictions on the additions allowed:
+//   - No change to CIDR; more work is required to determine how to handle.
+//   - No change to ProviderId nor ProviderNetworkID; these are immutable.
+func (s *Subnet) UpdateOps(args network.SubnetInfo) ([]txn.Op, error) {
```
Comment (nammn) [inline state/subnets.go]: Do we want to use this patch for updating the Subnet SpaceID?
If yes:
- need to change to support a move from a space other than `alpha` name -> what implication does that have for the existing code? It is used for the updating part from the provider update
-  Does a check for exiting key `FanLocalUnderlay` make sense for this patch?

Comment (nammn) [reply state/subnets.go]: @manadart  ^

File: apiserver/testing/stub_network.go [hunk 4]
```diff
@@ -5,6 +5,7 @@ package testing
 
 import (
 	"fmt"
+	"gopkg.in/mgo.v2/txn"
```
Comment (nammn) [inline apiserver/testing/stub_network.go]: own grouping

File: apiserver/facades/client/spaces/opfactory.go [hunk 5]
```diff
@@ -55,3 +59,13 @@ func (f *opFactory) NewRenameSpaceModelOp(fromName, toName string) (state.ModelO
 	}
 	return NewRenameSpaceModelOp(f.st.IsController(), controllerSettings, &renameSpaceStateShim{f.st}, space, toName), nil
 }
+
+// NewUpdateSpaceModelOp (OpFactory) returns an operation
+// for updating a space.
+func (f *opFactory) NewUpdateSpaceModelOp(spaceName string, subnets []networkingcommon.BackingSubnet) (MoveToSpaceModelOp, error) {
```
Comment (nammn) [inline apiserver/facades/client/spaces/opfactory.go]: rename to move-to-space from update

File: state/subnets.go (hunk 3 above)
Comment (manadart) [reply state/subnets.go]: Yes, we no longer restrict to moving from `alpha`.
And yes, we need to move fan overlay/underlay together.

Comment (manadart) [inline state/subnets.go]: You could break the space change ops out of this separately and just use those.

File: apiserver/common/networkingcommon/types.go [hunk 6]
```diff
@@ -36,6 +37,8 @@ type BackingSubnet interface {
 	Status() string
 	SpaceName() string
 	SpaceID() string
+	Refresh() error
```
Comment (manadart) [inline apiserver/common/networkingcommon/types.go]: I'd rather not add this here, opting instead for a tightly focussed interface at the usage site.

Comment (nammn) [reply apiserver/common/networkingcommon/types.go]: The latest patch d983087c9190e5e77b43920e4bb5f9d4f2c0c27b should fix that by using the needed Subnet interface defined at `move.go` itself.

File: state/subnets.go (hunk 3 above)
Comment (nammn) [reply state/subnets.go]: For now, I was planning to exclude the changes to that place, as the patch was getting bigger. 
d983087c9190e5e77b43920e4bb5f9d4f2c0c27b uses an extracted version of `getUpdateSpaceOps`, which only checks for the spaceName and omits the fan overlay/underlay check. Would that work? I expected that we don't need this, as we are not really touching the machines.

File: apiserver/testing/stub_network.go [hunk 7]
```diff
@@ -319,6 +320,14 @@ type FakeSubnet struct {
 
 var _ networkingcommon.BackingSubnet = (*FakeSubnet)(nil)
 
+func (f *FakeSubnet) UpdateOps(args network.SubnetInfo) ([]txn.Op, error) {
```
Comment (nammn) [inline apiserver/testing/stub_network.go]: can be removed

File: apiserver/facades/client/spaces/spaces_test.go [hunk 8]
```diff
@@ -250,6 +250,106 @@ func (s *SpaceTestMockSuite) TestRenameSpaceErrorProviderSpacesSupport(c *gc.C)
...
+	}}
+
+	obtainedResponse, err := s.api.MoveToSpace(args)
+	expectedErrMsg := "cannot move CIDR \"10.0.0.0/24\" from space \"alpha\" to space: \"myspace\", as this would violate the current application constraint: \"^myspace\" on application \"mysql\""
+
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(obtainedResponse.Results[0].Error.Message, gc.Equals, expectedErrMsg)
+}
+
+func (s *SpaceTestMockSuite) TestMoveToSpaceErrorProviderSpacesSupport(c *gc.C) {
```
Comment (nammn) [inline apiserver/facades/client/spaces/spaces_test.go]: need to add tests for the cases:

- CIDR wrong
- space not exist
- CIDR already in space 
- add force option for constraint and test

File: apiserver/facades/client/spaces/opfactory.go [hunk 9]
```diff
@@ -55,3 +58,9 @@ func (f *opFactory) NewRenameSpaceModelOp(fromName, toName string) (state.ModelO
 	}
 	return NewRenameSpaceModelOp(f.st.IsController(), controllerSettings, &renameSpaceStateShim{f.st}, space, toName), nil
 }
+
+// NewMoveToSpaceModelOp (OpFactory) returns an operation
+// for updating a space.
```
Comment (nammn) [inline apiserver/facades/client/spaces/opfactory.go]: ... to move a list of CIDRs to a space

File: apiserver/facades/client/spaces/remove.go [hunk 10]
```diff
@@ -24,18 +23,19 @@ type RemoveSpace interface {
 
 type spaceRemoveModelOp struct {
 	space   RemoveSpace
-	subnets []Subnet
+	subnets []RemoveSubnet
 }
 
-type Subnet interface {
+// RemoveSubnet describes a subnet which can be moved
```
Comment (nammn) [inline apiserver/facades/client/spaces/remove.go]: Other name?

File: apiserver/params/network.go [hunk 11]
```diff
@@ -931,6 +931,36 @@ type CreateSpaceParams struct {
 	ProviderId string   `json:"provider-id,omitempty"`
 }
 
+// MoveToSpacesParams holds the arguments of the MoveToSpace API call.
+type MoveToSpacesParams struct {
+	MoveToSpace []MoveToSpaceParams `json:"update-spaces"`
+}
+
+// MoveToSpaceParams holds the space tag and at least one subnet required to move subnet to another space
+type MoveToSpaceParams struct {
```
Comment (nammn) [inline apiserver/params/network.go]: check alternative naming

File: state/subnets.go [hunk 12]
```diff
@@ -237,6 +237,30 @@ func (s *Subnet) setSpace(subnets mongo.Collection) error {
 	return nil
 }
 
+func (s *Subnet) UpdateSpaceOps(spaceName string) ([]txn.Op, error) {
```
Comment (nammn) [inline state/subnets.go]: comment

File: state/subnets.go (hunk 3 above)
Comment (manadart) [reply state/subnets.go]: If you look at `state.AddSpace` you can see that we never explicitly indicate a fan overlay space. It always inherits the space from its underlay.
